lib/template-loader.php

```diff
@@ -42,34 +42,61 @@ function gutenberg_add_template_loader_filters() {
 
 /**
- * Filters into the "{$type}_template" hooks to record the current template hierarchy.
+ * Get the template hierarchy for a given template type.
  *
- * The method returns an empty result for every template so that a 'wp_template' post
- * is used instead.
+ * Internally, this filters into the "{$type}_template_hierarchy" hook to record the type-specific template hierarchy.
  *
- * @see gutenberg_template_include_filter
+ * @param string $template_type A template type.
+ * @return string[] A list of template candidates, in descending order of priority.
+ */
+function get_template_hierachy( $template_type ) {
+	$get_template_function     = 'get_' . str_replace( '-', '_', $template_type ) . '_template'; // front-page -> get_front_page_template
+	$template_hierarchy_filter = str_replace( '-', '', $template_type ) . '_template_hierarchy'; // front-page -> frontpage_template_hierarchy
+
+	$result                             = array();
+	$template_hierarchy_filter_function = function( $templates ) use ( &$result ) {
+		$result = $templates;
+		return $templates;
+	};
+
+	add_filter( $template_hierarchy_filter, $template_hierarchy_filter_function, 20, 1 );
+	call_user_func( $get_template_function ); // This invokes template_hierarchy_filter
+	remove_filter( $template_hierarchy_filter, $template_hierarchy_filter_function, 20 );
+
+	return $result;
+}
+
+/**
+ * Filters into the "{$type}_template" hooks to redirect them to the Full Site Editing template canvas.
+ *
+ * Internally, this communicates the block content that needs to be used by the template canvas through a global variable.
  *
  * @param string $template  Path to the template. See locate_template().
  * @param string $type      Sanitized filename without extension.
  * @param array  $templates A list of template candidates, in descending order of priority.
- * @return string Empty string to ensure template file is considered not found.
+ * @return string The path to the Full Site Editing template canvas file.
  */
 function gutenberg_override_query_template( $template, $type, array $templates = array() ) {
-	global $_wp_current_template_hierarchy;
+	global $_wp_current_template_id, $_wp_current_template_content;
 
-	if ( ! is_array( $_wp_current_template_hierarchy ) ) {
-		$_wp_current_template_hierarchy = $templates;
-	} else {
-		$_wp_current_template_hierarchy = array_merge( $_wp_current_template_hierarchy, $templates );
+	$current_template_post = gutenberg_find_template_post( $templates );
+
+	if ( $current_template_post ) {
+		$_wp_current_template_id      = $current_template_post->ID;
+		$_wp_current_template_content = empty( $current_template_post->post_content ) ? __( 'Empty template.', 'gutenberg' ) : $current_template_post->post_content;
 	}
 
-	return '';
+	// Add extra hooks for template canvas.
+	add_action( 'wp_head', 'gutenberg_viewport_meta_tag', 0 );
+	remove_action( 'wp_head', '_wp_render_title_tag', 1 );
```

Comment:
I know this was here before, but it's not immediately clear from the comment why this is being removed.

Reply:
aeb7a04